src/hotspot/cpu/s390/stubGenerator_s390.cpp

```diff
@@ -2857,6 +2858,48 @@ class StubGenerator: public StubCodeGenerator {
     return start;
   }
 
+  address generate_nmethod_entry_barrier() {
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "nmethod_entry_barrier");
+
+    address start = __ pc();
+
+    // Save caller's sp & return_pc
+    __ push_frame(frame::z_abi_16_size);
+    __ z_stmg(Z_R14, Z_R15, _z_abi16(callers_sp), Z_SP);
+
+    // Prep arg for VM call
+    // Since we call address* -> int and not address -> int, we must provide an extra level of indirection.
+    // We construct a pointer to the location of R14 stored above.
+    __ z_xgr(Z_R2, Z_R2);
+    __ z_ag(Z_R2, _z_abi(return_pc), 0, Z_SP);
+
+    // Call BarrierSetNMethod::nmethod_stub_entry_barrier(address* return_address_ptr)
+    __ push_frame_abi160(0);
+    __ call_VM_leaf(CAST_FROM_FN_PTR(address, BarrierSetNMethod::nmethod_stub_entry_barrier), Z_R2);
+    __ pop_frame();
+
+    // Restore caller's sp & return_pc
+    __ z_lmg(Z_R14, Z_R15, _z_abi(callers_sp), Z_SP);
+    __ pop_frame();
+
+    // Check return val of vm call
+    // if (return val != 0)
+    // return to caller
+    __ z_cfi(Z_R2, 0);
+    __ z_bcr(Assembler::bcondNotEqual, Z_R14);
+
```

Comment:
`z_ltr(Z_R2, Z_R2);`
is the preferred way of testing register contents. I would then use the "speaking" alias
`z_brnz(Z_R14);
Note: there are subtle semantic differences between "not equal" and "not zero". See assembler_s390.hpp.

Reply:
I changed to the preferred test instruction. However, it seems like z_brnz won't work in this situation because the branch address is in a register and not a label.

Side note it too me a second to realize that I am using bcr and brnz is an alias for brc. It seems odd that there isn't a "speaking" alias for bcr (or maybe I'm just not finding it).

Reply:
You are right. There are no "speaking" aliases for brc. "Branch on Condition" is not used frequently in VM code. Therefore, nobody felt enough pain so far to define the aliases.
May I please request you use the condition `Assembler::bcondNotZero`? `LTR` compares the register value against zero.

Reply:
I'm happy to change bcondNotEqual -> bcondNotZero, especially since it seems to more clearly represent the intent. Thanks for the suggestion.

Out of curiosity: How are they different? You mentioned above that there are semantic differences between not equal and not zero, but in assembler_s390.hpp, it looks like bcondNotZero is actually an alias for bcondNotEqual.

Reply:
You are right. The two conditions are technically identical. There is a semantic difference. If you say "NotEqual", you imply that you compared two (arbitrary) values. If you say "NotZero", it is clear that you tested the value against zero. When you use LTR to test a value and copy it at the same time, it may be confusing for the not so profound s390 hacker. Example:
```
z_ltr(Z_R1, Z_R2);
z_bcr(Assembler::bcondNotEqual, R14);

```Was the contents of the registers compared before r2 was copied into r1? For sure not, you say. But that's because you are a profound s390 hacker meanwhile. :-)

Reply:
I see the reasoning here; thanks for clarifying.

Thanks also for your kind words. I'm not sure that I quite at the level of s390 hacker indicated above, but I'm working on it :-)